src/robot/running/userkeywordrunner.py

```diff
@@ -235,7 +235,19 @@ def _resolve_arguments(self, args, variables=None):
         self.arguments.resolve(args, variables)
         if not variables:
             return []
-        return [(n, variables.replace_scalar(v)) for n, v in self.embedded_args]
+        return list(self._replace(variables))
+
+    def _replace(self, variables):
+        for n, v in self.embedded_args:
+            try:
+                v = eval(v)
+                if is_dict_like(v):
+                    yield n, DotDict(v)
+                elif is_list_like(v):
+                    yield n, variables.replace_list(v)
+            except:
+                pass
+            yield n, variables.replace_scalar(v)
```

Comment:
Blindly using eval and especially empty `except:` are both pretty ugly. Also, wouldn't this code turn embedded argument values that can be evaluated in Python to the resulting value? For example, arguments like `True` and `1 + 2` (both strings) would be converted to Python `True` and integer `3`.

Evaluation should *only* be done if the original argument name starts with `@` or `&`. Then it should be validated that if it starts with `@`, the result should be a list, and if it starts with `&` it should be a dict. Also, evaluation should rather be done using `ast.literal_eval` than `eval`.

It would probably be a good idea to implement dedicated utility functions for converting string to lists and dicts. In the beginning these utils could just use `ast.literal_eval` (and validate the result is list/dict), but in the future we could try coming up with other, more readable formats as discussed in #2699.